src/mpxmanager.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.16.
.TH MPXMANAGER.SH "1" "October 2020" "mpxmanager.sh 1.2.0" "User Commands"
.SH NAME
mpxmanager.sh \- manual page for mpxmanager.sh 1.2.0
.SH SYNOPSIS
.B mpxmanager.sh
[\fI\,ACTION\/\fR]
.SH DESCRIPTION
MPX aware window manager
.SS "Action:"
.TP
\fB\-\-hash\fR
Displays git hash
.TP
\fB\-\-help\fR, \fB\-h\fR
Print this help
.TP
\fB\-\-mode\fR, \fB\-m\fR
run with a predefined settings. Valid options are: xmousecontrol
.TP
\fB\-\-name\fR
Name of executable
.TP
\fB\-\-path\fR
Path to the config file
.TP
\fB\-\-quit\fR
Exists a running instance
.TP
\fB\-\-recompile\fR
Recompile with program. Needed for changes to config to be loaded
.TP
\fB\-\-restart\fR
Restart the program
.TP
\fB\-\-send=\fR<name=value>, \fB\-s\fR
send argument(s) to running instance (can be called multiple times)
.TP
\fB\-\-set=\fR<name=value>
set argument(s) for new instances (can be called multiple times)
.TP
\fB\-v\fR, \fB\-\-version\fR
Print script information
.PP
Modes:
Each mode is a non\-WM feature
.IP
xmousecontrol enables control of the mouse with the keyboard
mpx\-start   loads mpx config file, creates masters and attaches slaves to the appropriate masters
mpx\-stop    removes all non\-default masters
mpx\-restart mpx\-stop and mpx\-start
mpx\-restore   Like mpx\-start but does not create any new masters
mpx\-split   takes all slaves that are sending events and add them to a newly created master. To make the timeout longer, change the value of POLL_COUNT/POLL_INTERVAL via \fB\-\-set\fR
.IP
EXIT STATUS
.IP
The following exit values shall be returned:
.TP
0
Successful completion.
.TP
1
Display related error liking not being able to find the XServer or an X command failing
.TP
64
System call failed
.TP
65
Another WM is already running
.TP
66
WM is not responding to a request
.TP
67
An invalid option was specified
.TP
68
Validation failed
.SH EXAMPLES
.TP
mpxmanager.sh
#start the window manager
.TP
mpxmanager.sh \-\-set="VAR=VALUE"
#start the window manager and set VAR to VALUE
.TP
mpxmanager.sh \-\-send="VAR=VALUE"
#set VAR to VALUE for a running instance
.PP
Usage: ${SCRIPT_NAME} [ACTION]
MPX aware window manager
.SS "Action:"
.TP
\-\-hash
Displays git hash
.TP
\-\-help, \-h
Print this help
.TP
\-\-mode, \-m
run with a predefined settings. Valid options are: xmousecontrol
.TP
\-\-name
Name of executable
.TP
\-\-path
Path to the config file
.TP
\-\-quit
Exists a running instance
.TP
\-\-recompile
Recompile with program. Needed for changes to config to be loaded
.TP
\-\-restart
Restart the program
.TP
\-\-send=<name=value>, \-s
send argument(s) to running instance (can be called multiple times)
.TP
\-\-set=<name=value>
set argument(s) for new instances (can be called multiple times)
.TP
\-v, \-\-version
Print script information
.PP
Modes:
Each mode is a non\-WM feature
.IP
xmousecontrol enables control of the mouse with the keyboard
mpx\-start   loads mpx config file, creates masters and attaches slaves to the appropriate masters
mpx\-stop    removes all non\-default masters
mpx\-restart mpx\-stop and mpx\-start
mpx\-restore   Like mpx\-start but does not create any new masters
mpx\-split   takes all slaves that are sending events and add them to a newly created master. To make the timeout longer, change the value of POLL_COUNT/POLL_INTERVAL via \-\-set
.IP
EXIT STATUS
.IP
The following exit values shall be returned:
.TP
0
Successful completion.
.TP
1
Display related error liking not being able to find the XServer or an X command failing
.TP
64
System call failed
.TP
65
Another WM is already running
.TP
66
WM is not responding to a request
.TP
67
An invalid option was specified
.TP
68
Validation failed
.TP
${SCRIPT_NAME}
#start the window manager
.TP
${SCRIPT_NAME} \-\-set="VAR=VALUE"
#start the window manager and set VAR to VALUE
.TP
${SCRIPT_NAME} \-\-send="VAR=VALUE"
#set VAR to VALUE for a running instance


src/mpxmanager.sh
#!/bin/bash
#================================================================
# HEADER
#================================================================
#%Usage: ${SCRIPT_NAME} [ACTION]
#%MPX aware window manager
#%
#%
#%Action:
#%    --recompile                   Recompile with program. Needed for changes to config to be loaded
#%    --restart                     Restart the program
#%    --name                        Name of executable
#%    --path                        Path to the config file
#%    --send=<name=value>, -s       send argument(s) to running instance (can be called multiple times)
#%    --set=<name=value>            set argument(s) for new instances (can be called multiple times)
#%    --mode, -m                    run with a predefined settings. Valid options are: xmousecontrol
#%    -h, --help                    Print this help
#%    --quit                        Exists a running instance
#%    -v, --version                 Print script information
#%
#%Modes:
#%Each mode is a non-WM feature
#%      xmousecontrol enables control of the mouse with the keyboard
#%      mpx-start   loads mpx config file, creates masters and attaches slaves to the appropriate masters
#%      mpx-stop    removes all non-default masters
#%      mpx-restart mpx-stop and mpx-start
#%      mpx-restore   Like mpx-start but does not create any new masters
#%      mpx-split   takes all slaves that are sending events and add them to a newly created master. To make the timeout longer, change the value of POLL_COUNT/POLL_INTERVAL via --set
#%
#%Examples:
#%    ${SCRIPT_NAME}              #start the window manager
#%    ${SCRIPT_NAME} --set="VAR=VALUE"             #start the window manager and set VAR to VALUE
#%    ${SCRIPT_NAME} --send="VAR=VALUE"             #set VAR to VALUE for a running instance
#%
#================================================================
#- IMPLEMENTATION
#-    version         ${SCRIPT_NAME} (taaparthur.no-ip.org) 0.9.1
#-    author          Arthur Williams
#-    license         MIT
#================================================================
# END_OF_HEADER
#================================================================
#MAN generated with help2man -No mpxmanager.1 ./mpxmanager.sh

set -e

displayHelp(){
    SCRIPT_HEADSIZE=$(head -200 ${0} |grep -n "^# END_OF_HEADER" | cut -f1 -d:)
    SCRIPT_NAME="$(basename ${0})"
    head -${SCRIPT_HEADSIZE:-99} ${0} | grep -e "^#[%+]" | sed -e "s/^#[%+-]//g" -e "s/\${SCRIPT_NAME}/${SCRIPT_NAME}/g" ;
}

run(){
    exec $path/$fileName $@
}
displayVersion(){
    echo "0.9.8"
}

fileName=$(echo "mpxmanager-"$(sed "s/ /-/g" -- < <(uname -sm)))
path=$HOME/.mpxmanager
optspec=":vhs-:"

LIBS="-lpthread -lm -lX11 -lXi -lxcb -lxcb-xinput -lxcb-xtest -lxcb-ewmh -lxcb-icccm -lxcb-randr -lX11-xcb"

getopts "$optspec" optchar || run
    case "${optchar}" in
        -)
            case "${OPTARG}" in
                force-restart)
                    pid=$(pgrep $fileName)
                    kill $pid;
                    while kill -0 $pid; do
                        sleep 1
                    done
                    shift
                    run $@
                    ;;
                restart)
                    run "--send=restart"
                    ;;
                recompile)
                    mkdir -p $path
                    [ -f $path/config.cpp ] || cp /usr/share/mpxmanager/sample-config.cpp $path/config.cpp
                    shift
                    gcc $path/config.cpp -o $path/$fileName -lmpxmanager $LIBS "$@"
                    ;;
                help)
                    displayHelp
                    ;;
                version)
                    displayVersion
                    ;;
                name)
                    echo $fileName
                    ;;
                path)
                    echo $path
                    ;;
                quit)
                    run "--send=quit"
                    ;;
                *)
                    run $@
                    ;;
            esac;;
        v)
            displayVersion
            ;;
        h)
            displayHelp
            ;;
        *)
            run $@
            ;;
    esac
